### ISSUE TITLE
Remove 'runserver' in favor of 'start'

### ISSUE THREAD

**Issue by kdmccormick** [overhangio/2u-tutor-adoption#26]
## Context

There are two ways to bring up services in dev mode:
1. `tutor dev runserver myservice`
2. `tutor dev start [myservice1 myservice2 ...]`

Command 1 has the advantage of accepting volume mounts and allowing debugger attachment.

Command 2 has the advantage of:
* accepting multiple services, or
* accepting no services, in which case all enabled services are started. This is helpful in that it doesn't "leave out" any important services, such as the MFE container. [Relevant forum thread](https://discuss.overhang.io/t/mfe-not-running-by-default-on-development-start-documentation-error/2578).

Are there other substantive differences? Could they be reconciled into a single command that supports all features?

### Update

We have already:
* added the ability for `tutor [dev|local] start` to mount volumes using the new `--mount` command
* allowed debugger attachment for containers started by `tutor dev start`
* marked `tutor dev runserver` as deprecated, since it no longer has any advantages over `tutor dev start`. This was done in: https://github.com/overhangio/2u-tutor-adoption/issues/61

## Acceptance Criteria

* In order to give folks time to stop using `runserver`, wait until Tutor v14 is released.
* Go through [official plugins](https://overhang.io/tutor/plugins). For each plugin:
  * Ensure that the plugin is ported to the V1 API
  * Ensure that the plugin hooks into the `COMPOSE_MOUNTS` filter in order to automatically bind-mount folders as appropriate.
  * Replace any references to `runserver` with `start`
  * Replace any references to the `--volume` option with an equivalent usage of the `--mount` option.
* Remove the implementation of `tutor dev runserver`.

**Comment by regisb** [comment]
> Are there other substantive differences? 

- `runserver` allocates a tty, which makes it possible to `breakpoint()` python code.
- containers run as `runserver` are no longer reachable by containers run as `start` -- and this is a problem. For instance, with `tutor dev runserver lms` and `tutor dev start -d cms` then we can no longer login to the Studio, because the CMS makes an internal call to the LMS.

We can certainly do a better job at better documenting these commands, but I'm afraid that we need both.

**Comment by kdmccormick** [comment]
@regisb Is there any reason we wouldn't want `start` to allocate a tty like [devstack's dev.up does](https://github.com/openedx/devstack/blob/3a6e515d2cfbe76a2e53742fb03733ac1ea5c457/docker-compose.yml#L423-L424)?

**Comment by regisb** [comment]
This is a very good point! I attempted this some time ago and it didn't work, though I can't remember why. I tried again just now and I now realise that I can `docker attach` to any container that was started with `tty/stdin_open: true`, thus allowing me to troubleshoot any container without the use of `docker-compose run`.
Thus, yes we can get rid of `docker-compose run` for long-running services, such as `runserver`. This is fantastic news, as it allows us to greatly simplify the CLI :partying_face:

**Comment by kdmccormick** [comment]
@regisb Awesome!

I'm excited that we can drop `tutor dev runserver`. Would you still want `tutor dev run <service> <cmd>` to exist for running one-off commands? For example: tests, migrations, management commands, and so on.

Or should those workflows be replaced with `tutor dev start <service> && tutor dev exec <service> <cmd>`?

**Comment by regisb** [comment]
> Would you still want tutor dev run <service> <cmd> to exist for running one-off commands? For example: tests, migrations, management commands, and so on.

Yes, `dev run` commands would still be useful. For instance I use the following commands very frequently:

- `tutor dev run lms bash`
- `tutor dev run --no-deps lms ./manage.py lms shell`
- `tutor local run ecommerce ./manage.py shell -c "from django.contrib.auth import get_user_model; get_user_model().objects.filter(email='admin@overhang.io').update(is_staff=True, is_superuser=True)"`

For one-off tasks, `exec` is still inferior to `run` because it ignores the environment variables defined in the docker-compose*.yml files.

**Comment by kdmccormick** [comment]
> Yes, dev run commands would still be useful. 

I agree. Since switching from devstack, I do appreciate how Tutor can run commands with spinning up a persistent container.

> exec is still inferior to run because it ignores the environment variables defined in the docker-compose*.yml files.

I have not been able to recreate this. I tried adding an environment variable to docker-compose.yml:
```
$ vi $(tutor config printroot)/env/local/docker-compose.yml
    ...
    lms:
      image: docker.io/overhangio/openedx:13.1.5-nightly
      environment:
        SERVICE_VARIANT: lms
        UWSGI_WORKERS: 2
        SETTINGS: ${TUTOR_EDX_PLATFORM_SETTINGS:-tutor.production}
        MY_VAR: my_val  # <- added this line
    ...
$ tutor dev run lms printenv MY_VAR
my_val
$ tutor dev start -d lms
$ tutor exec lms bash
app@f4960a487696:~/edx-platform$ echo $MY_VAR
my_val
app@f4960a487696:~/edx-platform$
```
and found that both `exec` and `run` pick up on the new environment variable.

I'll update the acceptance criteria of this issue to keep `run` in the CLI because I think it's useful, but I'm still curious to understand what's going on with environment variables.

**Comment by regisb** [comment]
> and found that both exec and run pick up on the new environment variable.

Right. I guess I was wrong about that. Now that you mention it, the actual problem that I faced in the past is that `exec` bypasses the Docker entrypoint. Sorry about the confusion. Tutor used to run images with heavy entrypoint scripts. This is mostly no longer the case, but it might still cause surprises from time to time.

**Comment by kdmccormick** [comment]
>  the actual problem that I faced in the past is that exec bypasses the Docker entrypoint.

Whew, that is a very confusing Docker nuance 😛 

That sheds light on one weird behavior I've noticed: using `tutor [local|k8s] exec`, Django management commands *almost* work perfectly, except that DJANGO_SETTINGS_MODULE isn't set. After I run `export DJANGO_SETTINGS_MODULE=[lms|cms].envs.tutor.production`, everything works fine.

I think that it's because the Dockerfile [sets DJANGO_SETTINGS_MODULE](https://github.com/overhangio/tutor/blob/3e2c4aebf40f9c2afd5711aa4b791cf2944df6bc/tutor/templates/build/openedx/Dockerfile#L207-L208) via an [entrypoint script](https://github.com/overhangio/tutor/blob/master/tutor/templates/build/openedx/bin/docker-entrypoint.sh) instead of setting it directly as using `ENV`. It seems like DJANGO_SETTINGS_MODULE is the only environment variable set this way.

Do you know of any reason we couldn't pull DJANGO_SETTINGS_MODULE into the Dockerfile?

**Comment by kdmccormick** [comment]
This is effectively blocked by #43 , because we can't remove `runserver` until we have a replacement strategy for mounting local code.

**Comment by kdmccormick** [comment]
@regisb I updated the acceptance criteria of this ticket, can you take a look to see if you agree?

**Comment by regisb** [comment]
> I updated the acceptance criteria of this ticket, can you take a look to see if you agree?

I do! Sorry about the delay :-/

**Comment by kdmccormick** [comment]
No worries, thanks!

@Carlos-Muniz If/when you need a new ticket, this one is ready to go.

**Comment by Carlos-Muniz** [comment]
@kdmccormick I'm going to tackle this ticket if you want to assign it to me.

**Comment by Carlos-Muniz** [comment]
Of the official Tutor Plugins:
- Changes were made:
  - tutor-ecommerce: https://github.com/overhangio/tutor-ecommerce/pull/31
- Did not require any changes:
  - tutor-android
  - tutor-discovery
  - tutor-forum
  - tutor-mfe
  - tutor-minio
  - tutor-notes
  - tutor-webui
  - tutor-xqueue
- Source code is not available:
  - tutor-cairn
  - tutor-machine
  - tutor-monitor

**Comment by kdmccormick** [comment]
I think a few of these services should add callbacks to the `COMPOSE_MOUNTS` filter in order to enable local repo auto-mounting:
* tutor-xqueue: `--mount xqueue` should auto-mount to /openedx/xqueue in the xqueue and xqueue-job services.
* tutor-forum `--mount cs_comments_service` should auto-mount to /app/cs_comments_service in the forum and forum-job services.
* tutor-notes `--mount edx-notes-api` should auto-mount /app/edx-notes-api in the notes and notes-job services.

I agree about the rest of the services.

**Comment by Carlos-Muniz** [comment]
Local repo auto-mounting has been enabled for the following repos as requested:
- tutor-forum
- tutor-notes
- tutor-xqueue

**Comment by kdmccormick** [comment]
New removal PR, on top of nightly instead of master: https://github.com/overhangio/tutor/pull/718
